File: server/src/atlas_argus/domain/report.py
"""Report-section usability. Mirrors ``src/domain/report.ts``.

Priority: an unresolved conflict blocks outright; privileged material must be
withheld; anything unreviewed/excluded needs editorial attention; preserved
conflicts are usable only with the conflict disclosed.
"""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass(frozen=True)
class CitedClaim:
    """The slice of a claim that section impact depends on."""

    id: str
    status: str
    report_eligibility: str
    unresolved_conflict_summaries: tuple[str, ...] = ()


@dataclass(frozen=True)
class SectionImpact:
    status: str
    offending_claim_ids: list[str] = field(default_factory=list)
    note: str = ""


def compute_section_impact(paragraph_ref: str, cited: list[CitedClaim]) -> SectionImpact:
    blocked = [c for c in cited if c.report_eligibility == "blocked_by_conflict"]
    if blocked:
        conflict_names: list[str] = []
        for claim in blocked:
            for summary in claim.unresolved_conflict_summaries:
                if summary not in conflict_names:
                    conflict_names.append(summary)
        what = (
            "a disputed claim" if len(blocked) == 1 else f"{len(blocked)} disputed claims"
        )
        return SectionImpact(
            status="blocked",
            offending_claim_ids=[c.id for c in blocked],
            note=(
                f"Report blocked: paragraph {paragraph_ref} references {what} "
                f"({'; '.join(conflict_names)})."
            ),
        )

    privileged = [c for c in cited if c.report_eligibility == "privileged"]
    if privileged:
        return SectionImpact(
            status="privileged_material",
            offending_claim_ids=[c.id for c in privileged],
            note=(
                f"Paragraph {paragraph_ref} cites privileged material that must be "
                "withheld or replaced with a non-privileged source before production."
            ),
        )

    attention = [
        c for c in cited if c.report_eligibility in ("needs_review", "excluded")
    ]
    if attention:
        excluded = [c for c in attention if c.report_eligibility == "excluded"]
        if excluded:
            note = (
                f"Paragraph {paragraph_ref} cites {len(excluded)} excluded claim(s); "
                "the citation must be removed or replaced, and remaining support "
                "re-reviewed."
            )
        else:
            note = (
                f"Paragraph {paragraph_ref} cites claims that still need review "
                "before the section can be used."
            )
        return SectionImpact(
            status="needs_review",
            offending_claim_ids=[c.id for c in attention],
            note=note,
        )

    preserved = [c for c in cited if c.status == "preserved"]
    if preserved:
        return SectionImpact(
            status="eligible_with_disclosure",
            offending_claim_ids=[c.id for c in preserved],
            note=(
                f"Paragraph {paragraph_ref} is usable only if the preserved conflict "
                "is disclosed alongside both claims."
            ),
        )

    return SectionImpact(
        status="eligible",
        note="All cited claims are report-eligible with source citations intact.",
    )

File: src/domain/report.ts
import type { Claim, Conflict, ReportSection } from "./types";

export type SectionStatus =
  | "eligible"
  | "eligible_with_disclosure"
  | "blocked"
  | "privileged_material"
  | "needs_review";

export interface SectionImpact {
  status: SectionStatus;
  /** Claims that cause the blocking / warning state. */
  offendingClaimIds: string[];
  note: string;
}

/**
 * Derive a report section's usability from the claims it cites.
 * Priority: an unresolved conflict blocks outright; privileged material must
 * be withheld; anything unreviewed/excluded needs editorial attention;
 * preserved conflicts are usable only with the conflict disclosed.
 */
export function computeSectionImpact(
  section: Pick<ReportSection, "claimIds" | "paragraphRef">,
  claimsById: ReadonlyMap<string, Claim>,
  conflictsById: ReadonlyMap<string, Conflict>,
): SectionImpact {
  const cited = section.claimIds
    .map((id) => claimsById.get(id))
    .filter((c): c is Claim => c !== undefined);

  const blocked = cited.filter((c) => c.reportEligibility === "blocked_by_conflict");
  if (blocked.length > 0) {
    const conflictNames = new Set(
      blocked.flatMap((c) =>
        c.relatedConflictIds
          .map((id) => conflictsById.get(id))
          .filter((cf): cf is Conflict => cf !== undefined && cf.status === "unresolved")
          .map((cf) => cf.summary),
      ),
    );
    return {
      status: "blocked",
      offendingClaimIds: blocked.map((c) => c.id),
      note: `Report blocked: paragraph ${section.paragraphRef} references ${
        blocked.length === 1 ? "a disputed claim" : `${blocked.length} disputed claims`
      } (${[...conflictNames].join("; ")}).`,
    };
  }

  const privileged = cited.filter((c) => c.reportEligibility === "privileged");
  if (privileged.length > 0) {
    return {
      status: "privileged_material",
      offendingClaimIds: privileged.map((c) => c.id),
      note: `Paragraph ${section.paragraphRef} cites privileged material that must be withheld or replaced with a non-privileged source before production.`,
    };
  }

  const attention = cited.filter(
    (c) => c.reportEligibility === "needs_review" || c.reportEligibility === "excluded",
  );
  if (attention.length > 0) {
    const excluded = attention.filter((c) => c.reportEligibility === "excluded");
    return {
      status: "needs_review",
      offendingClaimIds: attention.map((c) => c.id),
      note:
        excluded.length > 0
          ? `Paragraph ${section.paragraphRef} cites ${excluded.length} excluded claim(s); the citation must be removed or replaced, and remaining support re-reviewed.`
          : `Paragraph ${section.paragraphRef} cites claims that still need review before the section can be used.`,
    };
  }

  const preserved = cited.filter((c) => c.status === "preserved");
  if (preserved.length > 0) {
    return {
      status: "eligible_with_disclosure",
      offendingClaimIds: preserved.map((c) => c.id),
      note: `Paragraph ${section.paragraphRef} is usable only if the preserved conflict is disclosed alongside both claims.`,
    };
  }

  return {
    status: "eligible",
    offendingClaimIds: [],
    note: `All cited claims are report-eligible with source citations intact.`,
  };
}

/** Report sections citing a given claim, with their current impact. */
export function sectionsCitingClaim(
  claimId: string,
  sections: readonly ReportSection[],
): ReportSection[] {
  return sections.filter((s) => s.claimIds.includes(claimId));
}
